python/ray/train/callbacks/results_preprocessors/aggregate_fn.py

```diff
@@ -0,0 +1,85 @@
+import logging
+from typing import Dict, List, Union, Tuple, Optional
+
+import numpy as np
+
+VALID_AGGREGATE_TYPES: Tuple[type] = (
+    int,
+    float,
+    np.float32,
+    np.float64,
+    np.int32,
+    np.int64,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class AggregateFn:
+    """An abstract class for aggregation function."""
+
+    def __call__(self):
+        raise NotImplementedError
+
+    def prepare(self, *args, **kwargs):
+        pass
+
+
+class Average(AggregateFn):
+    def __call__(self, values: List[Union[VALID_AGGREGATE_TYPES]]):
+        return np.nanmean(values)
+
+    def __repr__(self):
+        return "Average:"
+
+
+class Max(AggregateFn):
+    def __call__(self, values: List[Union[VALID_AGGREGATE_TYPES]]):
+        return np.nanmax(values)
+
+    def __repr__(self):
+        return "Max:"
+
+
+class WeightedAverage(AggregateFn):
+    def __init__(self, weight_key: Optional[str] = None):
+        self.weight_key = "" if weight_key is None else weight_key
+        self.weights = None
+
+    def __call__(self, values: List[Union[VALID_AGGREGATE_TYPES]]):
+        return np.nanmean(np.array(values) * self.weights / np.nansum(self.weights))
+
+    def __repr__(self):
+        return "Weighted average:"
+
+    def prepare(self, results: List[Dict]):
+        reported_metrics = set(results[0].keys())
+        self._get_weights(self.weight_key, reported_metrics, results)
```

Comment:
Oh I see this is similar to the pattern in `aggregate.py`. Maybe this can just import that function and call it? And then add additional logging + set `self.weights` directly in this method.

Reply:
Now it is all handled by a util method [`_get_values_from_results`](https://github.com/ray-project/ray/pull/22099/files#diff-023a1f7fbcaa536a9a4c0628378abd7991d5f0950eb495b6778ca7dcfa427e43R18).